### DISCUSSION TITLE
How to deserialize manually (C#)

### DISCUSSION THREAD

**Discussion by PeterWone** [Farfetch/kafkaflow#384] [Q&A]
KafkaFlow doesn't play well with our plugin system. For whatever reason, producers work but consumers don't handle messages even though middleware registered in the same statement does handle the messages.

Never mind why, I'm handling messages directly and it works. But other systems use KafkaFlow so object serializations are polluted with the type annotations that KafkaFlow presumably uses to dynamically reference types.

```json
    "actions": {
        "$type": "REDACTED.CouponAction[], REDACTED.ExternalModels",
        "$values": [
            {
                "coupon": "c1b16b81-3cde-4128-b521-0ac2f0815d26"
            }
        ]
    },
    "resources": {
        "$type": "REDACTED.Coupon[], REDACTED.ExternalModels",
        "$values": [
            {
                "rn": "c1b16b81-3cde-4128-b521-0ac2f0815d26",
                "channel": "Web",
                "currency": "AFN",
                "bets": {
                    "$type": "REDACTED.Bet[], REDACTED.ExternalModels",
                    "$values": [
                        {
                            "requestedStake": 0.0,
                            "formula": "doubles",
```

I can't just stuff that into `System.Text.JsonSerializer.Deserialize` and expect an object graph to emerge.

So I tried to use the KafkaFlow deserialiser...

```csharp
using KafkaFlow;
using KafkaFlow.Serializer;

var serctx = new SerializerContext("a.topic");
var ser = new JsonCoreSerializer();
var stream = File.Open("sample.json", FileMode.Open, FileAccess.Read);

var foo = await ser.DeserializeAsync(stream, typeof(Coupon), serctx);
```

which doesn't barf, but also doesn't work. The object is created in the right type but property values are not set. StackOverflow is completely useless, it doesn't have a single question mentioning KafkaFlow.Serializer.

I really hope someone familiar with the internal operation of KafkaFlow can advise me on a straightforward way to deserialise this format.

**Comment by gsferreira** [comment]
Hi @PeterWone 

Can you share the configuration of the Consumer/Producer?
The `.AddMiddlewares(middlewares => middleware` section.

Are you using the `KafkaFlow.Serializer.JsonCore` NuGet package?

**Comment by PeterWone** [comment]
The point of the exercise is to avoid using KafkaFlow altogether - it doesn't play well with the effects of our plugin system on dependency injection. But we are talking to other systems that use KafkaFlow, so I must deserialise what KafkaFlow serialises.

The actual problem was the presence of type information, as described in the question above, and I have since discovered that `Newtonsoft.Json` can deserialise this format (I was trying to use `System.Json.Text.JsonSerializer`).

So I think that's the answer to my question - "use `Newtonsoft.Json` for this format`.

If you guys know how to get `System.Json.Text.JsonSerializer` to parse embedded type information that would be awesome.

To answer your question, yes I was trying to use the `KafkaFlow.Serializer.JsonCore` NuGet package.
